### ISSUE TITLE
How can I use encrypted/hashed user credentials with the BasicAuthProvider?

### ISSUE THREAD

**Issue by EHJ-52n** [FraunhoferIOSB/FROST-Server#1426]
## Description

I want to use hashed passwords as outlined in the [documentation](https://fraunhoferiosb.github.io/FROST-Server/settings/auth.html).

## Steps to Reproduce

1. Create a `docker-compose.yml` with the following content:

   <details>
   <summary>docker-compose.yml</summary>

   ```yaml
   version: '3'

   name: 'frost'

   services:
     web:
       image: fraunhoferiosb/frost-server-http:2.0.4
       ports:
         - 8080:8080
       depends_on:
         - database
         - mosquitto
       environment:
         - serviceRootUrl=http://localhost:8080/FROST-Server
         - queueLoggingInterval=1000
         - logSensitiveData=true #disable on production!
         - plugins.multiDatastream.enable=true
         - http_cors_enable=true
         - http_cors_allowed_origins=*
         - bus_mqttBroker=tcp://mosquitto:1883
         - bus_sendQueueSize=2000
         - bus_sendWorkerPoolSize=10
         - bus_maxInFlight=2000
         - persistence_db_driver=org.postgresql.Driver
         - persistence_db_url=jdbc:postgresql://database:5432/sensorthings
         - persistence_db_username=sensorthings
         - persistence_db_password=ChangeMe
         - persistence_autoUpdateDatabase=true
         - auth_allowAnonymousRead=true
         - auth_db_autoUpdateDatabase=false
         - auth_autoUpdateDatabase=false
         - auth_db_driver=org.postgresql.Driver
         - auth_db_password=ChangeMe
         - auth_db_url=jdbc:postgresql://database:5432/sensorthings
         - auth_db_username=sensorthings
         - auth_plainTextPassword=false
         - auth_db_plainTextPassword=false
         - auth_provider=de.fraunhofer.iosb.ilt.frostserver.auth.basic.BasicAuthProvider

     mosquitto:
       image: eclipse-mosquitto
       volumes:
         - "./mosquitto.conf:/mosquitto/config/mosquitto.conf:ro"

     database:
       image: postgis/postgis:14-3.2-alpine
       environment:
         - POSTGRES_DB=sensorthings
         - POSTGRES_USER=sensorthings
         - POSTGRES_PASSWORD=ChangeMe
       volumes:
         - postgis_volume:/var/lib/postgresql/data

     pgadmin:
       image: dpage/pgadmin4
       environment:
         - PGADMIN_DEFAULT_EMAIL=b@c.de
         - PGADMIN_DEFAULT_PASSWORD=a
         - PGADMIN_DISABLE_POSTFIX=true
       ports:
         - 80:80

   volumes:
       postgis_volume:
   ```

   </details>

1. Add the `mosquitto.conf` with the following content:

   ```ini
   listener 1883
   allow_anonymous true
   ```

1. Start the set-up using:

   ```shell
   docker compose up -d ; docker compose logs -f web database
   ```

1. Trigger database creation from another terminal:

   ```shell
   curl -v -H "Accept: application/json" http://localhost:8080/FROST-Server/v1.1/ | jq
   ```

1. Access pgadmin at <http://localhost/> and log-in with the credentials `b@c.de` with password `a`.

1. Connect to the frost server database with the following configuration:

   * **Host**: `database`
   * **Username**: `sensorthings`
   * **Password**: `ChangeMe`

   (Somehow, the server ignores the setting `auth_autoUpdateDatabase=false` or `auth_db_autoUpdateDatabase=false` or both. That might be another issue.)

1. Connect to the database `sensorthings` and open the querytool ('Tools' → ' Querty Tool').

1. Insert the following SQL script:

   <details>
   <summary>SQL script</summary>

   ```sql
   DROP TABLE IF EXISTS public."USER_ROLES";
   DROP TABLE IF EXISTS public."USERS";

   -- Table: public.USERS
   CREATE TABLE IF NOT EXISTS public."USERS"
   (
       "USER_NAME" character varying(25) COLLATE pg_catalog."default" NOT NULL,
       "USER_PASS" character varying(255) COLLATE pg_catalog."default",
       CONSTRAINT "USERS_PKEY" PRIMARY KEY ("USER_NAME")
   )
   TABLESPACE pg_default;
   ALTER TABLE IF EXISTS public."USERS"
       OWNER to sensorthings;

   -- Table: public.USER_ROLES
   CREATE TABLE IF NOT EXISTS public."USER_ROLES"
   (
       "USER_NAME" character varying(25) COLLATE pg_catalog."default" NOT NULL,
       "ROLE_NAME" character varying(15) COLLATE pg_catalog."default" NOT NULL,
       CONSTRAINT "USER_ROLES_pkey" PRIMARY KEY ("USER_NAME", "ROLE_NAME"),
       CONSTRAINT "USER_ROLES_USERS_FKEY" FOREIGN KEY ("USER_NAME")
           REFERENCES public."USERS" ("USER_NAME") MATCH SIMPLE
           ON UPDATE CASCADE
           ON DELETE CASCADE
   )
   TABLESPACE pg_default;
   ALTER TABLE IF EXISTS public."USER_ROLES"
       OWNER to sensorthings;

   -- Insert our user
   CREATE EXTENSION IF NOT EXISTS pgcrypto;
   INSERT INTO "USERS" VALUES ('my_user', crypt('my_password', gen_salt('bf')));
   INSERT INTO "USER_ROLES" VALUES ('my_user', 'create');
   INSERT INTO "USER_ROLES" VALUES ('my_user', 'read');
   INSERT INTO "USER_ROLES" VALUES ('my_user', 'update');
   ```

   </details>

   The script deletes the tables `USERS` and `USER_ROLES`, recreates them and adds one user with hashed password and the roles `create`, `read`, and `update`.

1. You can verify that the password is hashed correctly by executing the following SQL statement and receive the result `true`.

   ```sql
   UPDATE public."USERS" SET "USER_PASS" = 'my_password' WHERE "USER_NAME" = 'my_user';
   ```

1. Try to create a thing:

   1. Create `thing.json` with the following content:

      <details>
      <summary>A minimal thing</summary>

      ```json
      {
        "@iot.id": "1",
        "name": "Weather Station",
        "description": "This is a weather station at an office building.",
        "properties": {
          "owner": "Some Company"
        }
      }
      ```

      </details>

   1. POST the thin against the things endpoint:

      ```shell
      curl -v -H "Accept: application/json" -H "Content-Type: application/json" -d @thing.json -u "my_user:my_password" http://localhost:8080/FROST-Server/v1.1/Things
      ```

      The result is a HTTP 401 error with the message:

      ```plain
      The request has not been applied because it lacks valid authentication credentials for the target resource.
      ```

1. If the password of the user is changed to plaintext using this sql statement:

   ```sql
   UPDATE public."USERS" SET "USER_PASS" = 'my_password' WHERE "USER_NAME" = 'my_user';
   ```

   the previous request works:

   <details>
   <summary>request and response</summary>

   ```shell
   $ curl -v -H "Accept: application/json" -H "Content-Type: application/json" -d @thing.json -u "my_user:my_password" http://localhost:8080/FROST-Server/v1.1/Things
   *   Trying 127.0.0.1:8080...
   * Connected to localhost (127.0.0.1) port 8080 (#0)
   * Server auth using Basic with user 'my_user'
   > POST /FROST-Server/v1.1/Things HTTP/1.1
   > Host: localhost:8080
   > Authorization: Basic bXlfdXNlcjpteV9wYXNzd29yZA==
   > User-Agent: curl/7.81.0
   > Accept: application/json
   > Content-Type: application/json
   > Content-Length: 204
   >
   * Mark bundle as not supporting multiuse
   < HTTP/1.1 201
   < Access-Control-Allow-Origin: *
   < Access-Control-Expose-Headers: Location
   < Location: http://localhost:8080/FROST-Server/v1.1/Things(3)
   < Content-Type: application/json;charset=UTF-8
   < Content-Length: 0
   < Date: Fri, 10 Feb 2023 15:51:28 GMT
   <
   * Connection #0 to host localhost left intact
   ```

   </details>

1. You can request the added resource without credentials as configured and required:

   <details>
   <summary>request and response</summary>

   ```shell
   curl -X GET -H "Accept: application/json" http://localhost:8080/FROST-Server/v1.1/Things\(1\) | jq
     % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                    Dload  Upload   Total   Spent    Left  Speed
   100   647  100   647    0     0   175k      0 --:--:-- --:--:-- --:--:--  210k
   {
     "@iot.selfLink": "http://localhost:8080/FROST-Server/v1.1/Things(1)",
     "@iot.id": 1,
     "name": "Weather Station",
     "description": "This is a weather station at an office building.",
     "properties": {
       "owner": "Some Company"
     },
     "Locations@iot.navigationLink": "http://localhost:8080/FROST-Server/v1.1/Things(1)/Locations",
     "HistoricalLocations@iot.navigationLink": "http://localhost:8080/FROST-Server/v1.1/Things(1)/HistoricalLocations",
     "Datastreams@iot.navigationLink": "http://localhost:8080/FROST-Server/v1.1/Things(1)/Datastreams",
     "MultiDatastreams@iot.navigationLink": "http://localhost:8080/FROST-Server/v1.1/Things(1)/MultiDatastreams"
   }
   ```

   </details>

## Expected Behaviour

I would like to be able to use hashed passwords, hence, please tell me how to configure the server and set the user passwords.
It did not work when following the [documentation on auth](https://fraunhoferiosb.github.io/FROST-Server/settings/auth.html) and [pg_crypto](https://www.postgresql.org/docs/14/pgcrypto.html).

**Comment by hylkevds** [comment]
The FROST version you are using in your docker-compose file is 2.0.4, but encrypted passwords are a 2.1.x feature.

**Comment by EHJ-52n** [comment]
To put it in other words, this is currently not supported.

I assume this because I cannot find any [tag > 2.0.6](https://github.com/FraunhoferIOSB/FROST-Server/tags?before=v2.0.6).
Do you have any information, when this will be implemented?
You might add a note [in the documentation](https://github.com/FraunhoferIOSB/FROST-Server/blob/v2.x/docs/settings/auth.md?plain=1#L91-L95), that this feature is/will be availabe in version ≥ 2.1.0

**Comment by hylkevds** [comment]
I've released 2.1.0.
Adding version tags to the configuration options is a good idea.
